FutureFeatures.md

```diff
@@ -66,16 +66,160 @@ Options under consideration:
+specified independently of JS and the Web platform but allow a natural
+integration.
+
+### Opaque reference types
+
+The first sub-feature is to extend [module imports](Modules.md#imports-and-exports)
+to allow modules to import *opaque reference types*. "Opaque" means that the
+reference type itself has no structural content and does not, e.g., define any
+methods or fields. Once imported, an opaque reference type can be used in the
+signature of other imported functions. Thus, the point of an opaque reference
+type is to be passed to and returned from exported functions.
+
+Reference types are allowed to be used as the types of locals, parameters
+and return types. Additionally, references would be allowed as operands to
+operators that treat their values as black boxes (`conditional`, `comma`, maybe
+`eq`, etc.). A new `dynamic_cast` operator would be added to allow checked
+casting from any opaque reference type to any other opaque reference type.
+Whether the cast succeeds is up to the host environment; WebAssembly itself
+will define no a priori subtyping relationship.
+
+For reasons of safety and limiting nondeterminism, imported opaque reference
+types would not be able to be loaded from or stored to linear memory where they
+could otherwise be arbitrarily aliased as integers. Instead, a new set of
+operations would be added for allocating, deallocating, loading and storing
+from integer-indexed cells that could hold references and were not aliasable by
+linear memory. There are several important alternatives to consider:
+* LIFO allocation of cells versus (or in addition to) ad hoc explicit
+  allocation and deallocation.
+* Untyped cells (requiring `dynamic_cast` after load before use) vs. 
+  typed cells (with a separate index per type).
+* Cells that hold a weak reference.
+
+With opaque reference types expressed as imports, host environments can provide
+access to various kinds of reference-counted or garbage-collected host-defined
+objects via builtin modules. While this design does not mandate a JS VM or
+browser, it does allow natural integration with both
+[JS](FutureFeatures.md#js-integration) and
+[WebIDL](FutureFeatures.md#webidl-integration)
+in a Web environment.
+
+### JS integration
+
+Using [opaque reference types](FutureFeatures.md#opaque-reference-types),
+JS values could be made accessible to WebAssembly code through a builtin
+`js` module providing:
+* an exported `string` opaque reference type and exported functions
+  to allocate, query length, and index `string` values;
+* an exported `object` opaque reference type and exported functions
+  that correspond with the ES5 meta-object protocol;
+* an exported `value` opaque reference type with exported functions for
+  constructing `value`s from integers, floats, `object`s, and `string`s and
+  with exported functions for querying the type of a `value` and extracting the
+  abovementioned payload types.
+
```

Comment:
Well, according to ```typeof``` JS functions are not objects. :-) They do behave like objects other than being callable, though. When I see 'string' and 'object' listed but not 'function' I assume functions can't be stored as opaque references, because it's almost-mirroring-but-not-quite the JS type system. There will be more typeof enumerants once value types and typed objects land, I think?

It's probably sufficient to clarify this by stating what the three opaque reference types are, somehow:
```string``` stores JS string values (not ```String``` instances, which are objects)
```object``` stores JS objects (including the ```String``` wrapper), but not values
```value``` stores anything that can be stored in a JS variable (```string```, ```number```, ```object```...)